homeassistant/helpers/device.py

```diff
@@ -0,0 +1,75 @@
+"""Provides useful helpers for handling devices."""
+
+from homeassistant.core import HomeAssistant, callback
+
+from . import device_registry as dr, entity_registry as er
+
+
+@callback
+def async_entity_id_to_device_id(
+    hass: HomeAssistant,
+    entity_id_or_uuid: str,
+) -> str | None:
+    """Resolve the device id to the entity id or entity uuid."""
+
+    ent_reg = er.async_get(hass)
+
+    entity_id = er.async_validate_entity_id(ent_reg, entity_id_or_uuid)
+    if (entity := ent_reg.async_get(entity_id)) is None:
+        return None
+
+    return entity.device_id
+
+
+@callback
+def async_device_info_to_link_from_entity(
+    hass: HomeAssistant,
+    entity_id_or_uuid: str,
+) -> dr.DeviceInfo | None:
+    """DeviceInfo with information to link a device to a configuration entry in the link category from a entity id or entity uuid."""
+
+    dev_reg = dr.async_get(hass)
+
+    if (device_id := async_entity_id_to_device_id(hass, entity_id_or_uuid)) is None or (
+        device := dev_reg.async_get(device_id=device_id)
+    ) is None:
+        return None
+
+    return dr.DeviceInfo(
+        identifiers=device.identifiers,
+        connections=device.connections,
+    )
+
+
+@callback
+def async_remove_stale_devices_links_keep_entity_device(
+    hass: HomeAssistant,
+    entry_id: str,
+    source_entity_id_or_uuid: str,
+) -> None:
+    """Remove the link between stales devices and a configuration entry, keeping only the device that the informed entity is linked to."""
```

Comment:
Please limit docstring lines to max 72 characters in accordance with PEP8. The first line is a header and should be a single sentence. After that an optional body may follow separated from the header by a blank line.

Reply:
#120615